File: examples/ruby/spec/waits/waits_spec.rb
# frozen_string_literal: true

require 'spec_helper'

RSpec.describe 'Waits' do
  let(:driver) { start_session }

  it 'fails' do
    driver.get 'https://www.selenium.dev/selenium/web/dynamic.html'
    driver.find_element(id: 'adder').click

    expect {
      driver.find_element(id: 'box0')
    }.to raise_error(Selenium::WebDriver::Error::NoSuchElementError)
  end

  it 'sleeps' do
    driver.get 'https://www.selenium.dev/selenium/web/dynamic.html'
    driver.find_element(id: 'adder').click

    sleep 1
    added = driver.find_element(id: 'box0')

    expect(added.dom_attribute(:class)).to eq('redbox')
  end

  it 'implicit' do
    driver.manage.timeouts.implicit_wait = 2
    driver.get 'https://www.selenium.dev/selenium/web/dynamic.html'
    driver.find_element(id: 'adder').click

    added = driver.find_element(id: 'box0')

    expect(added.dom_attribute(:class)).to eq('redbox')
  end

  it 'explicit' do
    driver.get 'https://www.selenium.dev/selenium/web/dynamic.html'
    revealed = driver.find_element(id: 'revealed')
    wait = Selenium::WebDriver::Wait.new

    driver.find_element(id: 'reveal').click
    wait.until { revealed.displayed? }

    revealed.send_keys('Displayed')
    expect(revealed.property(:value)).to eq('Displayed')
  end

  it 'options with explicit' do
    driver.get 'https://www.selenium.dev/selenium/web/dynamic.html'
    revealed = driver.find_element(id: 'revealed')
    errors = [Selenium::WebDriver::Error::NoSuchElementError,
              Selenium::WebDriver::Error::ElementNotInteractableError]
    wait = Selenium::WebDriver::Wait.new(timeout: 2,
                                         interval: 0.3,
                                         ignore: errors)

    driver.find_element(id: 'reveal').click
    wait.until { revealed.send_keys('Displayed') || true }

    expect(revealed.property(:value)).to eq('Displayed')
  end
end

File: examples/javascript/test/waits/waits.spec.js

const { By, Browser, until, Builder} = require('selenium-webdriver');
const assert = require("node:assert");


describe('Waits', function () {
    let driver;

    before(async function () {
        driver = new Builder()
          .forBrowser(Browser.CHROME)
          .build();
    });

    after(async () => await driver.quit());

    it('fail', async function () {
        await driver.get('https://www.selenium.dev/selenium/web/dynamic.html');
        await driver.findElement(By.id("adder")).click();

        await assert.rejects(async () => {
              await driver.findElement(By.id("box0"))
          },
          Error
        )
    });

    it('sleep', async function () {
        await driver.get('https://www.selenium.dev/selenium/web/dynamic.html');
        await driver.findElement(By.id("adder")).click();

        await driver.sleep(2000);
        let added = await driver.findElement(By.id("box0"));

        assert.equal(await added.getAttribute('class'), "redbox")
    });

    it('implicit', async function () {
        await driver.manage().setTimeouts({ implicit: 2000 });
        await driver.get('https://www.selenium.dev/selenium/web/dynamic.html');
        await driver.findElement(By.id("adder")).click();

        let added = await driver.findElement(By.id("box0"));

        assert.equal(await added.getAttribute('class'), "redbox")
    });

    it('explicit', async function () {
        await driver.get('https://www.selenium.dev/selenium/web/dynamic.html');
        let revealed = await driver.findElement(By.id("revealed"));
        await driver.findElement(By.id("reveal")).click();
        await driver.wait(until.elementIsVisible(revealed), 2000);
        await revealed.sendKeys("Displayed");
        assert.equal(await revealed.getAttribute("value"), "Displayed")
    })
});